### PR TITLE
[consensus/application] Async Verify Nits

### PR DESCRIPTION
**Pull request by patrick-ogrady**
Related: #2604

### CODE REVIEW

File: consensus/src/application/marshaled.rs [hunk 1]
```diff
@@ -196,24 +197,6 @@ where
                 let tx_closed = tx.closed();
                 pin_mut!(tx_closed);
 
-                // Helper to clean up verification context for this round
```
Comment (patrick-ogrady) [inline consensus/src/application/marshaled.rs:199]: Ah, I now understand what this was doing (if we attempt to build on a notarization it assumes its been certified)

Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:199]: Is it safe to remove all certifications less than this (if there are forks tho)?

File: consensus/src/marshal/actor.rs [hunk 2]
```diff
@@ -918,9 +926,13 @@ where
             panic!("failed to finalize: {e}");
         }
 
-        // Update metrics and send tip update to application
+        // Update metrics and send tip update to application.
+        // The tip must have a direct finalization (see get_latest() comment).
```
Comment (patrick-ogrady) [inline consensus/src/marshal/actor.rs]: nit: remove this line

File: consensus/src/marshal/actor.rs [hunk 3]
```diff
@@ -918,9 +926,13 @@ where
...
         if height > self.tip {
```
Comment (patrick-ogrady) [inline consensus/src/marshal/actor.rs]: a safer route may be to only enter if `round` is not none?

File: consensus/src/application/marshaled.rs [hunk 4]
```diff
@@ -539,7 +537,14 @@ where
         context: Context<Self::Digest, S::PublicKey>,
         digest: Self::Digest,
     ) -> oneshot::Receiver<bool> {
-        Self::store_verification_context(&self.verification_contexts, context, digest).await;
+        let finalized_round = self.last_finalized_round.lock().await.take();
```
Comment (patrick-ogrady) [inline consensus/src/application/marshaled.rs]: Why not just pass the Arc<Mutex<>> in?

File: consensus/src/application/marshaled.rs (hunk 1 above)
Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:199]: Yes, it's safe. Once a block at round R is finalized, all blocks on competing forks at rounds <= R are dead and will never be certified. We only need verification contexts for blocks that might still be certified, and those are all at rounds > the finalized round.

File: consensus/src/application/marshaled.rs [hunk 5]
```diff
@@ -106,6 +106,7 @@ where
     last_built: Arc<Mutex<Option<(Round, B)>>>,
     verification_contexts: Arc<Mutex<VerificationContexts<E, B, S>>>,
     certification_txs: Arc<Mutex<Vec<oneshot::Sender<bool>>>>,
```
Comment (patrick-ogrady) [inline consensus/src/application/marshaled.rs]: I don't think we actually need to store these (just dropping if we don't have the certification should be sufficient)?

File: consensus/src/application/marshaled.rs [hunk 6]
```diff
@@ -560,35 +534,26 @@ where
...
-            drop(txs_guard);
-
-            // If we don't have a verification context, we cannot verify the block.
-            // In this event, we return a receiver that will never resolve to signal
-            // to consensus that we should time out the view and vote to nullify.
+            // Verify is always called before certify for a given proposal, so if we
+            // don't have a verification context here, it means this proposal was never
+            // verified. Return a receiver that never resolves to signal to consensus
+            // that we should time out the view and vote to nullify.
+            let (_tx, rx) = oneshot::channel();
```
Comment (clabby) [inline consensus/src/application/marshaled.rs:556]: This won't hang like the previous code did, since we drop `tx` right away. https://play.rust-lang.org/?version=stable&mode=release&edition=2024&gist=e3cd9e2cf8a70334c9ee27740fa699ba

You'll end up entering this branch in the voter:
https://github.com/commonwarexyz/monorepo/blob/3375949367fd2fc64be4a9a635efc5e3be709b0e/consensus/src/simplex/actors/voter/actor.rs#L935-L943

rather than hitting the timeout:
https://github.com/commonwarexyz/monorepo/blob/3375949367fd2fc64be4a9a635efc5e3be709b0e/consensus/src/simplex/actors/voter/actor.rs#L851-L855

Comment (clabby) [reply consensus/src/application/marshaled.rs:556]: (Is this intentional? If so, docs here should be updated.)

Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:556]: Yes, intentional

File: consensus/src/application/marshaled.rs [hunk 7]
```diff
@@ -560,35 +534,26 @@ where
...
-    async fn certify(&mut self, payload: Self::Digest) -> oneshot::Receiver<bool> {
-        // Look up context by block digest and pop the oldest (by round) context from the map.
-        // Using pop_first ensures each concurrent certification for the same digest gets a
-        // unique context, which is critical when a block is re-proposed across multiple views.
-        //
-        // This works because the voter certifies views in ascending order: when the same block
-        // appears in multiple views (e.g., re-proposals at epoch boundaries), certify is called
-        // for the lower view first.
-        let mut contexts_guard = self.verification_contexts.lock().await;
+    async fn certify(&mut self, round: Round, payload: Self::Digest) -> oneshot::Receiver<bool> {
```
Comment (patrick-ogrady) [inline consensus/src/application/marshaled.rs:537]: Do we need both of these? I feel like `round` ~should be sufficient?

Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:537]: Ah, if you don't get verify you'll still want payload ...

File: consensus/src/marshal/mod.rs [hunk 8]
```diff
@@ -2031,6 +2041,161 @@ mod tests {
         })
     }
 
+    /// Regression test for cleanup bug in #2604.
```
Comment (patrick-ogrady) [inline consensus/src/marshal/mod.rs:2044]: nit: remove PR specific statements, just include details of edge case

File: consensus/src/application/marshaled.rs (hunk 7 above)
Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:537]: We should add a comment explaining as much in the consensus trait

Comment (patrick-ogrady) [reply consensus/src/application/marshaled.rs:537]: If we are taking the view that blocks may not contain any consensus data, this is a must.

If we instead assume the digest changes because it contains `Round`, then this is a no-op.
